### ISSUE TITLE
Crash In SDLRouterStatusProvider

### ISSUE THREAD

**Issue by martin-price87** [smartdevicelink/sdl_android#688]
This is coming through in my crash reports, it appears to happen quite regularly in the wild however i have not seen it during internal testing. Hope you guys are able to help. Let me know if you need more info.

```
Caused by java.lang.RuntimeException: Can't create handler inside thread that has not called Looper.prepare()
       at android.os.Handler.<init>(Handler.java:204)
       at android.os.Handler.<init>(Handler.java:118)
       at com.smartdevicelink.transport.SdlRouterStatusProvider$ClientHandler.<init>(SdlRouterStatusProvider.java:130)
       at com.smartdevicelink.transport.SdlRouterStatusProvider.<init>(SdlRouterStatusProvider.java:71)
       at com.smartdevicelink.transport.SdlBroadcastReceiver$4.onConnectionStatusUpdate(SdlBroadcastReceiver.java:373)
       at com.smartdevicelink.transport.SdlRouterStatusProvider.checkIsConnected(SdlRouterStatusProvider.java:80)
       at com.smartdevicelink.transport.SdlBroadcastReceiver$5.onListObtained(SdlBroadcastReceiver.java:399)
       at com.smartdevicelink.transport.RouterServiceValidator$2.httpCallComplete(RouterServiceValidator.java:470)
       at com.smartdevicelink.util.HttpRequestTask.doInBackground(HttpRequestTask.java:124)
       at com.smartdevicelink.util.HttpRequestTask.doInBackground(HttpRequestTask.java:16)
       at android.os.AsyncTask$2.call(AsyncTask.java:333)
       at java.util.concurrent.FutureTask.run(FutureTask.java:266)
       at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:245)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
       at java.lang.Thread.run(Thread.java:764)
```

**Comment by brettywhite** [comment]
Hi @martin-price87 are you able to provide any additional information or context. Essentially please provide what you can out of the issue template.

**Comment by martin-price87** [comment]
[uk.co.radioplayer_issue_356_crash_5A93E41B02780001104EB5786494767A_DNE_3_v2.txt](https://github.com/smartdevicelink/sdl_android/files/1758908/uk.co.radioplayer_issue_356_crash_5A93E41B02780001104EB5786494767A_DNE_3_v2.txt)

Hi brett, 

Thanks for getting back to me so quickly.

I'm not able to provide any more context than is in this attached crash report. I am not able to repoduce it on our internal builds.

Sorry I don't wish to be presumptuous but from what i can tell from the code, this is where the problem occurs :
```
	public SdlRouterStatusProvider(Context context, ComponentName service, ConnectedStatusCallback callback){
		if(context == null || service == null || callback == null){
			throw new IllegalStateException("Supplied params are not correct. Context == null? "+ (context==null) + " ComponentName == null? " + (service == null) + " ConnectedStatusListener == null? " + callback);
		}
		this.context = context;
		this.routerService = service;
		this.cb = callback;
		**# this.clientMessenger = new Messenger(new ClientHandler(this));**

	}
```
SdlRouterStatusProvider seems to be being instantiated off the main thread. The reason for this is not obvious to me, however would it be better to guard against this by ensuring SdlRouterStatusProvider instances are created on the main thread since each instance will create a  new handler?

```
new Handler(Looper.getMainLooper()) {
    @Override
    public void handleMessage(Message message) {
        // Create instances here.
    }
};

```
Thanks again 

Martin

**Comment by brettywhite** [comment]
Thanks, we'll take a look at it.

**Comment by Tnnnguyen** [comment]
Here is my observation:
The constructor of `SdlRouterStatusProvider` can be invoked from 
```SdlRouterBroadcast.requestTransportStatus``` which can be invoked from 
```SdlRouterBroadcast.queryForConnectedService```.

The call `SdlRouterBroadcast.queryForConnectedService` should be made from `MainActivity` (per SDL Android guide) but I have seen developers making this call from a Service (that is `SdlService.java`, still works by the way).  Moreover, there is no stopping the developers from making this call from a background thread which by default does not have a Looper associated with it. 

Base on the given exception log, I think this is what happened.  If the call is originated from one of the Android components (Activity, Service, BroadcastReceiver) running on the main thread, then the looper is automatically prepared and starts to loop as seen in the `main` method of `ActivityThread.java` [here](http://androidxref.com/8.0.0_r4/xref/frameworks/base/core/java/android/app/ActivityThread.java).

If the call is originated from any class running in a background thread, and if the thread has not initiated a looper, it cannot create a handler, resulting in the RuntimeException above because a handler will need a looper.

Furthermore, the Looper should not be a local Looper for the background thread but should be the main Looper because the thread can die and so does the Looper associated with it, at which point the Messenger/Handler objects may still be doing work which will create problem.

My suggestion for the fix is attach the main Looper to the `ClientHandler` object in the constructor:
```
static class ClientHandler extends Handler {
		final WeakReference<SdlRouterStatusProvider> provider;

		 public ClientHandler(SdlRouterStatusProvider provider){
		 		super(Looper.getMainLooper()); // NEW
			 	this.provider = new WeakReference<SdlRouterStatusProvider>(provider);
		 }
```
Since the main Looper is always initiated and running with the application start, combining with ClientHandler being static, memory leak shouldn't be a problem.
@joeygrover, what do you think? If it's ok I will create a fix for this.

**Comment by joeygrover** [comment]
@Tnnnguyen yes you are correct. The reason this exact case is happening is because in production, the library is attempting to grab a new trusted router service list (where during debug it might skip it) and thus creating an async task (`HttpRequestTask`) to connect to the backed. The callback of that task is executed on that async thread, so therefore, the handler is created on that same thread without a looper. I think the solution presented is efficient since the code being executed is very minimal so attaching it to the main looper should be fine. You have my 👍 to move forward.

**Comment by joeygrover** [comment]
Fixed with #697
